### DISCUSSION TITLE
How do I add Neumann conditions？

### DISCUSSION THREAD

**Discussion by Richard-happy** [libMesh/libmesh#3796] [Q&A] [answered]
Hello, I'm learning libmesh with examples which are showed on Web. I want to solve an possion with Neumann conditions,like
$$-∆u=f \ in\ Ω$$
$$\frac{\partial u}{\partial\mathbf{n}}=g\ on\ \partial\Omega$$.
I wonder if it's a good deal  that I  put boundary on the right side directly when coding.
$$\int_{\Omega}\nabla u\nabla\psi=\int_{\Omega}{\psi f\ }+\int_{\partial\Omega}\psi g$$
Hope for your answering.

**Comment by roystgnr** [comment, marked as answer]
Yes, substituting the Neumann boundary term into the weak formulation's boundary integral (after integration by parts) to get a forcing term, called "weak enforcement", is the standard way to impose a Neumann boundary.  IIRC we do this implicitly in a couple examples (which have zero-flux Neumann conditions, so the term just disappears), but we don't talk about it much since, unlike "strong enforcement" of Dirichlet conditions, the details of weak enforcement depend closely on your PDE and your weak formulation.

Actually, skimming through our examples ... it looks like we're weakly enforcing non-zero boundary fluxes in miscellaneous examples 3 and 9, reduced-basis example 7, systems-of-equations examples 4, 5, and 6 ... and yet not once did we put "Neumann" into a comment users could search for.  Sorry about that!

**Reply by roystgnr** [reply]
Oh, one caveat that may become important as you couple in more physics or as you debug your code:

Weak enforcement does only give you your Neumann condition *weakly* - the boundary flux of your approximate solution will converge towards the exact solution, but won't in general equal it.  This is a good thing from some perspectives (the flexibility pays off as lower L2 and H1 error norms, whereas trying to instead enforce Neumann BCs strongly can even break your error convergence rates) but it can be surprising to people new to finite elements.

**Reply by Richard-happy** [reply]
Thank you for your patience answer. I'll read the examples above.And thanks again for your caveat. :D
